File: scripts/monitoring/logger.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'time'
require 'optparse'

module Gmux
  module Scripts
    class Logger
      STORE_PATH = File.join(Dir.home, '.gmux', 'sessions.json')
      LOG_DIR = File.join(Dir.home, '.gmux', 'logs')

      def initialize(options = {})
        @interval = options[:interval] || 10
        @verbose = options[:verbose] || false
        @log_dir = options[:log_dir] || LOG_DIR
        @watch_session = options[:session]
        @rotate = options[:rotate] || false
        @max_size = options[:max_size] || 10 * 1024 * 1024  # 10MB
        @compress = options[:compress] || false
        @format = options[:format] || 'text'
      end

      def run
        puts "=== GMUX Session Logger ==="
        puts "Interval: #{@interval}s"
        puts "Log dir: #{@log_dir}"
        puts

        ensure_log_dir

        if @rotate
          rotate_logs
        else
          capture_logs
        end
      end

      private

      def ensure_log_dir
        FileUtils.mkdir_p(@log_dir)
      end

      def load_sessions
        return {} unless File.exist?(STORE_PATH)
        JSON.parse(File.read(STORE_PATH))
      rescue JSON::ParserError
        {}
      end

      def capture_logs
        puts "Capturing logs... (Ctrl+C to stop)"
        puts

        running = true
        trap("INT") { running = false }

        while running
          sessions = load_sessions
          sessions_to_watch = @watch_session ?
            sessions.select { |name, _| name == @watch_session } :
            sessions.select { |_, s| s['status'] == 'running' }

          if sessions_to_watch.empty?
            print "\r  No running sessions to capture..."
            sleep @interval
            next
          end

          sessions_to_watch.each do |name, session|
            capture_session_log(name, session)
          end

          sleep @interval
        end

        puts "\nStopped capturing."
      end

      def capture_session_log(session_name, session)
        pane_id = session['tmuxPaneId']
        return unless pane_id

        log_file = get_log_file(session_name)

        # Capture pane output
        output = `tmux capture-pane -t #{pane_id} -p -S -500 2>/dev/null`
        return unless $?.success?

        # Append to log file
        File.open(log_file, 'a') do |f|
          f.puts "=" * 60
          f.puts "Capture at: #{Time.now.iso8601}"
          f.puts "Session: #{session_name}"
          f.puts "Pane: #{pane_id}"
          f.puts "-" * 60
          f.puts output
          f.puts
        end

        puts "  Captured: #{session_name} -> #{File.basename(log_file)}" if @verbose
      end

      def get_log_file(session_name)
        timestamp = Time.now.strftime('%Y-%m-%d')
        extension = @format == 'json' ? 'json' : 'log'
        File.join(@log_dir, "#{session_name}_#{timestamp}.#{extension}")
      end

      def rotate_logs
        puts "Rotating logs..."
        puts

        log_files = Dir.glob(File.join(@log_dir, '*.log'))
        log_files.each do |log_file|
          size = File.size(log_file)
          next if size < @max_size

          puts "  Rotating: #{File.basename(log_file)} (#{format_size(size)})"

          # Create rotated file
          timestamp = Time.now.strftime('%Y%m%d_%H%M%S')
          rotated_file = "#{log_file}.#{timestamp}"

          if @compress
            # Compress and remove original
            system("gzip -c #{log_file} > #{rotated_file}.gz")
            File.delete(log_file)
            puts "    Compressed: #{File.basename(rotated_file)}.gz"
          else
            # Just rename
            File.rename(log_file, rotated_file)
            puts "    Renamed: #{File.basename(rotated_file)}"
          end
        end

        puts "Rotation complete."
      end

      def format_size(bytes)
        if bytes >= 1024 * 1024
          "#{(bytes / (1024.0 * 1024.0)).round(2)} MB"
        elsif bytes >= 1024
          "#{(bytes / 1024.0).round(2)} KB"
        else
          "#{bytes} B"
        end
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: #{$PROGRAM_NAME} [options]"

    opts.on('-i', '--interval SECONDS', Integer, "Capture interval in seconds (default: 10)") do |i|
      options[:interval] = i
    end

    opts.on('-s', '--session NAME', "Capture specific session") do |s|
      options[:session] = s
    end

    opts.on('-d', '--log-dir DIR', "Log directory (default: ~/.gmux/logs)") do |d|
      options[:log_dir] = d
    end

    opts.on('-r', '--rotate', "Rotate large log files") do
      options[:rotate] = true
    end

    opts.on('-m', '--max-size BYTES', Integer, "Max log size before rotation (default: 10MB)") do |m|
      options[:max_size] = m
    end

    opts.on('-z', '--compress', "Compress rotated logs") do
      options[:compress] = true
    end

    opts.on('-f', '--format FORMAT', %w[text json], "Log format (text, json)") do |f|
      options[:format] = f
    end

    opts.on('-v', '--verbose', "Show detailed output") do
      options[:verbose] = true
    end

    opts.on('-h', '--help', "Show this help") do
      puts opts
      exit
    end
  end.parse!

  Gmux::Scripts::Logger.new(options).run
end

File: scripts/monitoring/logger.py
#!/usr/bin/env python3
"""GMUX Session Logger - Capture and manage session logs."""

import gzip
import json
import logging
import os
import signal
import subprocess
import sys
import time
from argparse import ArgumentParser
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

STORE_PATH = Path.home() / ".gmux" / "sessions.json"
LOG_DIR = Path.home() / ".gmux" / "logs"


class Logger:
    def __init__(self, options: Dict[str, Any]):
        self.interval = options.get("interval", 10)
        self.verbose = options.get("verbose", False)
        self.log_dir = Path(options.get("log_dir", LOG_DIR))
        self.watch_session = options.get("session")
        self.rotate = options.get("rotate", False)
        self.max_size = options.get("max_size", 10 * 1024 * 1024)  # 10MB
        self.compress = options.get("compress", False)
        self.format = options.get("format", "text")
        self.running = True

        signal.signal(signal.SIGINT, self._handle_interrupt)

    def _handle_interrupt(self, signum, frame):
        self.running = False

    def run(self):
        print("=== GMUX Session Logger ===")
        print(f"Interval: {self.interval}s")
        print(f"Log dir: {self.log_dir}")
        print()

        self.ensure_log_dir()

        if self.rotate:
            self.rotate_logs()
        else:
            self.capture_logs()

    def ensure_log_dir(self):
        self.log_dir.mkdir(parents=True, exist_ok=True)

    def load_sessions(self) -> Dict[str, Any]:
        if not STORE_PATH.exists():
            return {}
        try:
            return json.loads(STORE_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            return {}

    def capture_logs(self):
        print("Capturing logs... (Ctrl+C to stop)")
        print()

        while self.running:
            sessions = self.load_sessions()
            sessions_to_watch = (
                {k: v for k, v in sessions.items() if k == self.watch_session}
                if self.watch_session
                else {k: v for k, v in sessions.items() if v.get("status") == "running"}
            )

            if not sessions_to_watch:
                print("\r  No running sessions to capture...", end="", flush=True)
                time.sleep(self.interval)
                continue

            for name, session in sessions_to_watch.items():
                self._capture_session_log(name, session)

            time.sleep(self.interval)

        print("\nStopped capturing.")

    def _capture_session_log(self, session_name: str, session: Dict[str, Any]):
        pane_id = session.get("tmuxPaneId")
        if not pane_id:
            return

        log_file = self._get_log_file(session_name)

        try:
            result = subprocess.run(
                ["tmux", "capture-pane", "-t", pane_id, "-p", "-S", "-500"],
                capture_output=True,
                text=True,
                timeout=5,
            )

            if result.returncode != 0:
                return

            output = result.stdout

            # Append to log file
            with open(log_file, "a") as f:
                f.write("=" * 60 + "\n")
                f.write(f"Capture at: {datetime.now().isoformat()}\n")
                f.write(f"Session: {session_name}\n")
                f.write(f"Pane: {pane_id}\n")
                f.write("-" * 60 + "\n")
                f.write(output)
                f.write("\n")

            if self.verbose:
                print(f"  Captured: {session_name} -> {log_file.name}")

        except (subprocess.TimeoutExpired, FileNotFoundError, OSError) as e:
            logger.debug(f"Capture error for {session_name}: {e}")

    def _get_log_file(self, session_name: str) -> Path:
        timestamp = datetime.now().strftime("%Y-%m-%d")
        extension = "json" if self.format == "json" else "log"
        return self.log_dir / f"{session_name}_{timestamp}.{extension}"

    def rotate_logs(self):
        print("Rotating logs...")
        print()

        log_files = sorted(self.log_dir.glob("*.log"))

        for log_file in log_files:
            size = log_file.stat().st_size

            if size < self.max_size:
                continue

            print(f"  Rotating: {log_file.name} ({self._format_size(size)})")

            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            rotated_file = log_file.with_name(f"{log_file.name}.{timestamp}")

            if self.compress:
                # Compress
                with open(log_file, "rb") as f_in:
                    with gzip.open(f"{rotated_file}.gz", "wb") as f_out:
                        f_out.writelines(f_in)

                log_file.unlink()
                print(f"    Compressed: {rotated_file.name}.gz")
            else:
                # Just rename
                log_file.rename(rotated_file)
                print(f"    Renamed: {rotated_file.name}")

        print("Rotation complete.")

    def _format_size(self, bytes: int) -> str:
        if bytes >= 1024 * 1024:
            return f"{bytes / (1024.0 * 1024.0):.2f} MB"
        elif bytes >= 1024:
            return f"{bytes / 1024.0:.2f} KB"
        else:
            return f"{bytes} B"


def main():
    parser = ArgumentParser(description="Capture and manage GMUX session logs")
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        default=10,
        help="Capture interval in seconds (default: 10)",
    )
    parser.add_argument("-s", "--session", help="Capture specific session")
    parser.add_argument(
        "-d",
        "--log-dir",
        dest="log_dir",
        default=str(LOG_DIR),
        help=f"Log directory (default: {LOG_DIR})",
    )
    parser.add_argument("-r", "--rotate", action="store_true", help="Rotate large log files")
    parser.add_argument(
        "-m",
        "--max-size",
        type=int,
        default=10 * 1024 * 1024,
        help="Max log size before rotation (default: 10485760)",
    )
    parser.add_argument("-z", "--compress", action="store_true", help="Compress rotated logs")
    parser.add_argument(
        "-f",
        "--format",
        choices=["text", "json"],
        default="text",
        help="Log format (default: text)",
    )
    parser.add_argument("-v", "--verbose", action="store_true", help="Show detailed output")

    args = parser.parse_args()

    logger_instance = Logger(vars(args))
    logger_instance.run()


if __name__ == "__main__":
    main()
